tests/TestCase/Controller/AppControllerTest.php

```diff
@@ -0,0 +1,437 @@
+        static::assertNotEmpty($this->App->{'Csrf'});
+        static::assertNotEmpty($this->App->{'Auth'});
+        static::assertNotEmpty($this->App->{'Modules'});
+        static::assertNotEmpty($this->App->{'Schema'});
+    }
+
+    /**
+     * test 'beforeFilter' not logged error
+     *
+     * @covers ::beforeFilter()
+     *
+     * @return void
+     */
+    public function testBeforeFilterLoginError() : void
+    {
+        $this->setupController();
+
+        $event = $this->App->dispatchEvent('Controller.initialize');
+        $this->App->beforeFilter($event);
+        $flash = $this->App->request->getSession()->read('Flash');
+
+        $expected = __('You are not logged or your session has expired, please provide login credentials');
+        $message = $flash['flash'][0]['message'];
+
+        static::assertEquals($expected, $message);
+    }
+
+    /**
+     * test 'beforeFilter' correct apiClient token setup
+     *
+     * @covers ::beforeFilter()
+     *
+     * @return void
+     */
+    public function testBeforeFilterCorrectTokens() : void
+    {
+        $expectedtokens = [];
+
+        $this->setupControllerAndLogin();
+
+        $expectedtokens = $this->App->Auth->user('tokens');
+
+        $event = $this->App->dispatchEvent('Controller.initialize');
+        $this->App->beforeFilter($event);
+
+        $apiClient = $this->accessProperty($this->App, 'apiClient');
+        $apiClientTokens = $this->accessProperty($apiClient, 'tokens');
+
+        static::assertEquals($expectedtokens, $apiClientTokens);
+    }
+
+    /**
+     * test SetupOutputTimezone
+     *
+     * @covers ::setupOutputTimezone
+     *
+     * @return void
+     */
+    public function testSetupOutputTimezone() : void
+    {
+        $this->setupController();
+        $expected = 'GMT';
+
+        // mock for AuthComponent
+        $mockedAuthComponent = $this->getMockBuilder('AuthComponent')
+                    ->setMethods(['user'])
+                    ->getMock();
+
+        // moch for user method
+        $mockedAuthComponent->method('user')
+            ->with('timezone')
+            ->willReturn($expected);
+
+        $this->App->Auth = $mockedAuthComponent;
+
+        $this->invokeMethod($this->App, 'setupOutputTimezone');
+
+        $configTimezone = Configure::read('I18n.timezone');
+
+        static::assertEquals($expected, $configTimezone);
+    }
+
+    /**
+     * Test `beforeRender` method
+     *
+     * @covers ::beforeRender()
+     * dataProvider prepareRequestProvider()
+     *
+     * @return void
+     */
+    public function testBeforeRender() : void
+    {
+        $user = $this->setupControllerAndLogin();
+
+        $event = $this->App->dispatchEvent('Controller.beforeRender');
+        $this->App->beforeRender($event);
+
+        static::assertArrayHasKey('user', $this->App->viewVars);
+        static::assertEquals($user, $this->App->viewVars['user']);
+    }
+
+    /**
+     * Test `beforeRender` method
```

Comment:
minor: this method is a different test, compared to the previous; some extra doc could be added.
I.e.:
```
Test `beforeRender` method, when updating tokens in session